### PR TITLE
[WIP] SameDiff: Batch-to-space and batch-to-space (nd)

### PR DESCRIPTION
**Pull request by maxpumperla**
SameDiff integration for these two ops. Tests still failing, must be doing something stupid. Following failure:

```
18:04:24.900 [main] INFO  o.n.a.gradcheck.GradCheckTransforms - batch to space failed on forward
18:04:24.902 [main] INFO  org.nd4j.autodiff.samediff.SameDiff - Executing op batch_to_space
18:04:24.902 [main] INFO  org.nd4j.autodiff.samediff.SameDiff -  Input shape for in is  [4, 1, 1, 1] Input shape for blocks is  [2, 1] Input shape for crops is  [2, 2] Output shape for batch_to_space is  [4, 1, 1, 1]
18:04:24.902 [main] INFO  org.nd4j.autodiff.samediff.SameDiff - Executing op mean
18:04:24.902 [main] INFO  org.nd4j.autodiff.samediff.SameDiff -  Input shape for batch_to_space_1 is  [4, 1, 1, 1] Output shape for loss is  []
18:04:24.912 [main] WARN  org.nd4j.autodiff.samediff.SameDiff - Action op state is null
org.nd4j.linalg.exception.ND4JIllegalStateException: No gradient found for batch_to_space
	at org.nd4j.autodiff.samediff.SameDiff$3.define(SameDiff.java:4183)
	at org.nd4j.autodiff.samediff.SameDiff.defineFunction(SameDiff.java:4055)
	at org.nd4j.autodiff.samediff.SameDiff.defineFunction(SameDiff.java:4039)
	at org.nd4j.autodiff.samediff.SameDiff.execBackwards(SameDiff.java:4115)
	at org.nd4j.autodiff.gradcheck.GradCheckUtil.checkGradients(GradCheckUtil.java:176)
	at org.nd4j.autodiff.gradcheck.GradCheckUtil.checkGradients(GradCheckUtil.java:137)
	at org.nd4j.autodiff.gradcheck.GradCheckUtil.checkGradients(GradCheckUtil.java:133)
	at org.nd4j.autodiff.gradcheck.GradCheckTransforms.testBatchToSpace(GradCheckTransforms.java:81)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

@AlexDBlack @agibsonccc if you have an idea what might go wrong here, please let me know. I'm still a bit new to this :)

### CODE REVIEW

**Comment by raver119** [comment]
nobody asked about gradient for these ops. put them to high-priority list as well please

**Comment by maxpumperla** [comment]
@raver119 these two ops are mutually inverse to each other, so we should be fine in this case. I'm pretty sure I'm just stupid here.

In general we do need gradients for everything. :)

**Comment by raver119** [comment]
Yes, they should be inverse. But who knows... Add them to list, and we'll check it and add separate impl if needed.

**Comment by AlexDBlack** [comment]
@maxpumperla I think this has a simpler explanation: you have 3 inputs (3 SDVariables) yet you return a single gradient. In general with samediff, if you have N input SDVariables, you have to return N gradients in doDiff.

More generally here (I suspect, but don't know for sure), 'blocks' and 'padding' should not actually be SDVariables? I suspect they should be treated as fixed values (and consequently should be int[], int[][] or INDArray or similar) and *not* SDVariables.

**Comment by agibsonccc** [comment]
@maxpumperla for the mapping, you *should* be able to use the properties based 1 liner I provided. Input parameters end up being properties/fields, not input args. @AlexDBlack is right that these should be ndarrays here instead. Make them properties and make it so we can use my 1 liner property mapper instead.

**Comment by maxpumperla** [comment]
@agibsonccc I've resolved shape issues I had for these ops with @raver119. What works is:

- Forward pass runs
- Backward pass for `batch_to_space` works in isolation (I run `doDiff` and feed a hand-made gradient), also works as intended.

Running gradient checks fails, because the input to `doDiff` of the `BatchToSpace` instance is `null`, i.e. the gradient isn't propagated as intended. I have no idea why. Do you have any pointers for me how to further debug this? @AlexDBlack maybe? 

(P.s.: I still have to take care of property mapping, I know, but note that this op is a little special)

**Comment by AlexDBlack** [comment]
> Running gradient checks fails, because the input to doDiff of the BatchToSpace instance is null, i.e. the gradient isn't propagated as intended. I have no idea why.

Yeah, not sure sorry. I'm not seeing a possible cause in your code. Happy to take a look in debugger though if you can't work it out.

**Comment by AlexDBlack** [comment]
So I think partly the problem here is all the extra stuff in the test, to check forward and backward results... my guess is those are impacting the SameDiff instance here.

Simplified version of the test, that does not give this exception:
```
    @Test
    public void testBatchToSpace() {
        Nd4j.getRandom().setSeed(1337);

        int miniBatch = 4;
        int[] inputShape = new int[]{miniBatch, 1, 1, 1};

        int M = 2;
        int[] blockShape = new int[]{M, 1};
        int[] cropShape = new int[]{M, 2};

        INDArray input = Nd4j.randn(inputShape);
        INDArray blocks = Nd4j.create(new float[]{2, 2}, blockShape);
        INDArray crops = Nd4j.create(new float[]{0,0,0,0}, cropShape);

        SameDiff sd = SameDiff.create();

        SDVariable sdInput = sd.var("in", inputShape);
        SDVariable t = sd.batchToSpace(sdInput, blocks, crops);
        SDVariable loss = sd.mean("loss", t);

        boolean ok;
        try{
            ok = GradCheckUtil.checkGradients(sd);
        } catch (Exception e){
            e.printStackTrace();
            ok = false;
        }

        assertTrue(ok);
    }
```

However, it (a) fails the gradient check (analytic gradients are all 0s), and (b) it takes an absurd amount of time to run (1 minute 34 sec for me)
image

image


Also @maxpumperla fyi one problem with using mean op for the loss is that the gradients are all the same at the output of the function (input to mean op)... which means you might not always pick up a bug (like, if the gradients are permuted somehow during backprop vs. how they should be, for a reshape op).
(We should probably change that in all of the other gradient checks, I realize some of the ones I implemented use this also... l2 norm might be a better choice?)

**Comment by raver119** [comment]
it takes 2 minutes with shape of [4, 1, 1, 1]???

**Comment by maxpumperla** [comment]
@raver119 yeah, I can confirm this behaviour. takes inexplicably long.

**Comment by raver119** [comment]
Simple question: what should i do to make this test running for 2 minutes? Yesterday and it was ok. Just checked today - still no issues.

**Comment by maxpumperla** [comment]
alright, will merge this. I'm adding more tests (e.g. for space_to_batch) and clean up a little more.
